src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -254,6 +276,550 @@ create_distributed_table(PG_FUNCTION_ARGS)
+								   bool shardCountIsStrict)
+{
+	/*
+	 * We disallow create_distributed_table_concurrently in transaction blocks
+	 * because we cannot handle preceding writes, and we block writes at the
+	 * very end of the operation so the transaction should end immediately after.
+	 */
+	PreventInTransactionBlock(true, "create_distributed_table_concurrently");
+
+	/* do not allow concurrent CreateDistributedTableConcurrently operations */
+	AcquireCreateDistributedTableConcurrentlyLock(relationId);
+
+	if (distributionMethod != DISTRIBUTE_BY_HASH)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("only hash-distributed tables can be distributed "
+							   "without blocking writes")));
+	}
+
+	if (ShardReplicationFactor > 1)
+	{
+		ereport(ERROR, (errmsg("cannot distribute a table concurrently when "
+							   "citus.shard_replication_factor > 1")));
+	}
+
+	EnsureCoordinatorIsInMetadata();
+	EnsureCitusTableCanBeCreated(relationId);
+
+	/*
+	 * Ensure table type is valid to be distributed. It should be either regular or citus local table.
+	 */
+	EnsureDistributableTable(relationId);
+
+	/*
+	 * we rely on citus_add_local_table_to_metadata, so it can generate irrelevant messages.
+	 * we want to error with a user friendly message if foreign keys are not supported.
+	 * We can miss foreign key violations because we are not holding locks, so relation
+	 * can be modified until we acquire the lock for the relation, but we do as much as we can
+	 * to be user friendly on foreign key violation messages.
+	 */
+
+	EnsureForeignKeysForDistributedTableConcurrently(relationId);
+
+	bool viaDeprecatedAPI = false;
+	char replicationModel = DecideReplicationModel(distributionMethod,
+												   colocateWithTableName,
+												   viaDeprecatedAPI);
+
+	/*
+	 * we fail transaction before local table conversion if the table could not be colocated with
+	 * given table. We should make those checks after local table conversion by acquiring locks to
+	 * the relation because the distribution column can be modified in that period.
+	 */
+	if (!IsColocateWithDefault(colocateWithTableName) && !IsColocateWithNone(
+			colocateWithTableName))
+	{
+		EnsureColocateWithTableIsValid(relationId, distributionMethod,
+									   distributionColumnName,
+									   colocateWithTableName);
+	}
+
+	/*
+	 * Get name of the table before possibly replacing it in
+	 * citus_add_local_table_to_metadata.
+	 */
+	char *tableName = get_rel_name(relationId);
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	RangeVar *rangeVar = makeRangeVar(schemaName, tableName, -1);
+
+	/* If table is a regular table, then we need to add it into metadata. */
+	if (!IsCitusTable(relationId))
+	{
+		/*
+		 * Before taking locks, convert the table into a Citus local table and commit
+		 * to allow shard split to see the shard.
+		 */
+		SendAddLocalTableToMetadataCommandOutsideTransaction(relationId);
```

Comment:
this is still a bit confusing:
```
select create_distributed_table_concurrently('test','x'), create_distributed_table_concurrently('test','x');
ERROR:  canceling the transaction since it was involved in a distributed deadlock
```
note that the second command implicitly works on the shard